src/utils/FixedQueue.ts

```diff
@@ -0,0 +1,52 @@
+import * as _ from 'underscore';
+import { deepClone } from '../../CloneUtils';
+
+export class FixedQueue<T> {
+
+  private queue: Array<T>;
+  private maxLength: number;
+
+  constructor(initialValues: Array<T> = [], maxLength: number = Number.MAX_VALUE) {
+    this.maxLength = maxLength;
+    this.queue = initialValues;
+    this.trimQueue();
+  }
+
+  immutablePush(object: any): FixedQueue<T> {
+    const queue: Array<T> = deepClone(this.queue);
+    queue.push(object);
```

Comment:
How is it the same reference? queue is a clone of this.queue so it should be another reference?